### ISSUE TITLE
DataWrapper and make_data_wrapper take shape arguments: does that make sense?

### ISSUE THREAD

**Issue by inducer** [inducer/pytato#46]
IIRC the intention was to use these to pass the same shape in symbolic form (if applicable) with the idea that the actual size would constrain (fix) the symbolic parameters. But I'm no longer sure this makes sense. After all, the constraining would happen as soon as the array "meets" a symbolically-shaped counterpart in (e.g.) arithmetic. Thoughts?

@xywei Do you remember if that 's accurate?

cc @kaushikcfd

**Comment by inducer** [comment]
(Discovered while chatting with @matthiasdiener)

**Comment by xywei** [comment]
There is a difference for code generation: using symbolic shape leads to parameterized kernels. For example,
```
import numpy as np
import pyopencl as cl
import pytato as pt
from pytato.array import SizeParam

ctx = cl.create_some_context()
queue = cl.CommandQueue(ctx)
data = np.arange(10).reshape([2, 5])

def print_code(arr):
    prog = pt.generate_loopy(arr, target=pt.LoopyPyOpenCLTarget(queue))
    print(prog.kernel)

# static shape
# { [_pt_out_dim0, _pt_out_dim1] : 0 <= _pt_out_dim0 <= 1 and 0 <= _pt_out_dim1 <= 4 }
ns1 = pt.Namespace()
x = pt.make_data_wrapper(ns1, data=data, name="x", shape=None)
print_code(x)

# dynamic shape
# [m, n] -> { [_pt_out_dim0, _pt_out_dim1] : 0 <= _pt_out_dim0 < m and 0 <= _pt_out_dim1 < n }
ns2 = pt.Namespace()
m = SizeParam(ns2, name="m")
n = SizeParam(ns2, name="n")
x = pt.make_data_wrapper(ns2, data=data, name="x", shape=("m", "n"))
print_code(x)
```

**Comment by kaushikcfd** [comment]
I agree with @xywei that the only benefit of size parameters in pytato is to make the DAG parametric in it, so that the cost of transforming the DAG is amortized over time for multiple expressions belonging to the parameter space. I.e. I'm fine with ignoring the slightly confusing syntax.

I'm also fine with deprecating `x = pt.make_data_wrapper(...)` in favor of:
```python
x = pt.make_placeholder(ns, shape="n", dtype=float)
pt.bind_data(x, np.random.rand(10))
```

Another possibility is define an "execution context" an object separate from pytato's namespace that keeps a record of the bound arguments and is passed to `BoundProgram`.

**Comment by inducer** [comment]
Let me rephrase @xywei's response to make sure I understand it:

What should happen in @xywei's second example is that we should generate code *as if* we didn't know the size of the data, so that when someone constructs a DAG that would generate the same kernels just with different sizes, we will get a cache hit.

If so, I like that interpretation, and I think we should adopt it. If you all agree (please state yes/no or vote), then what's left here is to document this intention.

As for @kaushikcfd's proposal to replace `DataWrapper` with placeholder binding: I like that, too, but that's #27 IMO, so let's continue that discussion there.
